File: apps/ios/Packages/UI/Sources/UI/media/TitleCard.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import DesignSystem
import Model
import SwiftUI

public struct TitleCard: View {
    @Environment(\.colorScheme) private var scheme
    private let title: TitleSummary
    private let poster: Image?
    private let onOpen: (MediaType, TMDBID) -> Void

    public init(title: TitleSummary, poster: Image? = nil, onOpen: @escaping (MediaType, TMDBID) -> Void) {
        self.title = title
        self.poster = poster
        self.onOpen = onOpen
    }

    public var body: some View {
        Button {
            onOpen(title.mediaType, title.id)
        } label: {
            VStack(alignment: .leading, spacing: 0) {
                Color.gaujaSurfaceVariant(scheme).aspectRatio(2 / 3, contentMode: .fit)
                    .overlay {
                        if let poster {
                            poster.resizable().scaledToFill().accessibilityHidden(true)
                        } else {
                            Text("No artwork").padding(12)
                        }
                    }.clipped()
                VStack(alignment: .leading, spacing: 4) {
                    Text(title.title.flatMap { $0.isEmpty ? nil : $0 } ?? String(localized: "Untitled"))
                        .font(.gaujaTitleMedium)
                    Text(mediaLabel).font(.gaujaBodySmall)
                    if let year = title.year { Text(year, format: .number.grouping(.never)).font(.gaujaBodySmall) }
                    if let rating = title.rating, rating.isFinite, (0...10).contains(rating) {
                        Text("Rating: \(rating, format: .number.precision(.fractionLength(1))) out of 10").font(
                            .gaujaBodySmall)
                    }
                    StatusBadge(status: title.status)
                    if let status = title.status4K { StatusBadge(status: status, is4K: true) }
                }.padding(12).frame(maxWidth: .infinity, alignment: .leading)
            }
            .foregroundStyle(Color.gaujaOnSurface(scheme))
            .background(Color.gaujaSurfaceContainer(scheme))
            .clipShape(RoundedRectangle(cornerRadius: 12))
            .contentShape(RoundedRectangle(cornerRadius: 12))
        }
        .buttonStyle(.plain).disabled(!canOpen)
        .accessibilityElement(children: .combine)
    }

    private var canOpen: Bool {
        switch title.mediaType {
        case .movie, .tv: true
        case .unrecognized: false
        }
    }
    private var mediaLabel: LocalizedStringKey {
        switch title.mediaType {
        case .movie: "Movie"
        case .tv: "TV"
        case .unrecognized: "Unknown media type"
        }
    }
}

#Preview("Title card") {
    if let id = TMDBID(rawValue: 42) {
        GaujaTheme {
            TitleCard(
                title: TitleSummary(
                    id: id, mediaType: .movie, title: "A journey along the river", year: 2026,
                    posterPath: nil, rating: 8.2, status: .available)
            ) { _, _ in }.frame(width: 220)
        }
    }
}
#Preview("Largest text", traits: .fixedLayout(width: 500, height: 1200)) {
    if let id = TMDBID(rawValue: 42) {
        GaujaTheme(scheme: .light) {
            TitleCard(
                title: TitleSummary(
                    id: id, mediaType: .tv, title: nil, year: nil,
                    posterPath: nil, rating: nil, status: .unrecognized(999))
            ) { _, _ in }.environment(\.dynamicTypeSize, .accessibility5)
        }
    }
}

File: apps/android/core/ui/src/main/kotlin/app/gauja/core/ui/media/TitleCard.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.ui.media

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.material3.Card
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.tooling.preview.PreviewScreenSizes
import androidx.compose.ui.unit.dp
import app.gauja.core.designsystem.GaujaTheme
import app.gauja.core.model.media.MediaType
import app.gauja.core.model.media.TitleSummary
import app.gauja.core.model.media.TmdbId
import app.gauja.core.model.status.MediaStatus
import app.gauja.core.ui.R

@Composable
fun TitleCard(
    title: TitleSummary,
    onOpen: (MediaType, TmdbId) -> Unit,
    modifier: Modifier = Modifier,
    poster: Painter? = null,
) {
    val knownType =
        when (title.mediaType) {
            MediaType.MOVIE,
            MediaType.TV -> true
            is MediaType.Unrecognized -> false
        }
    Card(
        onClick = { onOpen(title.mediaType, title.id) },
        modifier = modifier,
        enabled = knownType,
    ) {
        Column {
            Box(
                Modifier.fillMaxWidth()
                    .aspectRatio(2f / 3f)
                    .background(MaterialTheme.colorScheme.surfaceVariant),
                contentAlignment = Alignment.Center,
            ) {
                if (poster == null)
                    Text(stringResource(R.string.artwork_missing), Modifier.padding(12.dp))
                else
                    Image(
                        poster,
                        contentDescription = null,
                        modifier = Modifier.matchParentSize(),
                        contentScale = ContentScale.Crop,
                    )
            }
            Column(Modifier.padding(12.dp), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(
                    title.title?.takeIf(String::isNotBlank)
                        ?: stringResource(R.string.title_missing),
                    style = MaterialTheme.typography.titleMedium,
                )
                val type =
                    when (title.mediaType) {
                        MediaType.MOVIE -> R.string.media_movie
                        MediaType.TV -> R.string.media_tv
                        is MediaType.Unrecognized -> R.string.media_unknown
                    }
                Text(
                    listOfNotNull(stringResource(type), title.year?.toString()).joinToString(" · "),
                    style = MaterialTheme.typography.bodySmall,
                )
                title.rating
                    ?.takeIf { it.isFinite() && it in 0.0..10.0 }
                    ?.let {
                        Text(
                            stringResource(R.string.rating, it),
                            style = MaterialTheme.typography.bodySmall,
                        )
                    }
                StatusBadge(title.status)
                title.status4k?.let { StatusBadge(it, is4k = true) }
            }
        }
    }
}

@Preview
@PreviewScreenSizes
@Composable
private fun TitlePreview() {
    GaujaTheme {
        TitleCard(
            TitleSummary(
                TmdbId(42),
                MediaType.MOVIE,
                "A journey along the river",
                2026,
                null,
                8.2,
                MediaStatus.AVAILABLE,
            ),
            { _, _ -> },
            Modifier.widthIn(max = 220.dp),
        )
    }
}
